### PR TITLE
[flink] add option to support disable creating table in default database

### PR DESCRIPTION
**Pull request by legendtkl**
### Purpose

Linked issue: close https://github.com/apache/incubator-paimon/issues/1729

### Tests

FlinkCatalogTest

### CODE REVIEW

**Comment by zhuangchong** [comment]
I understand that it is supported in the paimon catalog, not just in the Flink engine. cc @FangYongs

**Comment by FangYongs** [comment]
@zhuangchong Agree, and @legendtkl could you move this to paimon catalog instead of flink catalog?

**Comment by legendtkl** [comment]
> @zhuangchong Agree, and @legendtkl could you move this to paimon catalog instead of flink catalog?

Got it. @FangYongs @zhuangchong

**Comment by legendtkl** [comment]
Hi, @FangYongs , @zhuangchong , I have addressed the review comment, would you help review it?

Thanks.

File: paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java [hunk 1]
```diff
@@ -205,4 +209,11 @@ public static Path databasePath(String warehouse, String database) {
     protected boolean isSystemDatabase(String database) {
         return SYSTEM_DATABASE_NAME.equals(database);
     }
+
+    protected boolean allowCreateTableInDefaultDB(Identifier identifier) {
+        return this.options
+                .getOptional(DISABLE_CREATE_TABLE_IN_DEFAULT_DB)
+                .map(val -> !(val & DEFAULT_DATABASE.equals(identifier.getDatabaseName())))
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:216]: Use `&&`？

Comment (FangYongs) [reply paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:216]: In addition, users can configure default database name with option `default-database` for flink

**Comment by FangYongs** [comment]
Thanks @legendtkl , left some comments

**Comment by legendtkl** [comment]
Hi, @FangYongs and @zhuangchong , let's have a discussion about the implementation.

1. Now flink engine support to specify the default database name with option `default-database` for flink, this will be stored in `org.apache.flink.table.catalog.AbstractCatalog`
2. But for paimon catalog, we can't access the default database specified by flink. If we check it in Paimon catalog, the info is missed.
3. So I propose to support this issue in FlinkCatalog now only. For Paimon default database support we can open a new ticket for it.

What do you think?

**Comment by FangYongs** [comment]
Thanks @legendtkl , it sounds good to me to support flink catalog in this PR

**Comment by FangYongs** [comment]
@legendtkl There're some conflicts

**Comment by legendtkl** [comment]
@FangYongs resolved, PTAL.

Thanks.

**Comment by FangYongs** [comment]
Thanks @legendtkl , +1

**Review by FangYongs — APPROVED** [review]

**Pull request merged**